### DISCUSSION TITLE
Saving checkbox values as they are checked or unchecked

### DISCUSSION THREAD

**Discussion by marcdegagne** [livewire/livewire#4751] [1. Help] [answered]
I have been trying to save checkbox values as they are checked by the user. No save button for this part is needed.

This is a many-to-many relationship and therefore, it’s the relation table that gets updated with the new values.

Since this a many-to-many relations, the ideal saving method is sync(), but the array returned by Livewire to updatedSelected() is the same as initial values from the database and that is where I am stuck.

The problem is when values exist in the DB, load the page, uncheck one or more and the array returned to updatedSelected() is the same as it was initially loaded on page load.

What I am expecting is only the values that are left checked to properly sync.
1) Is this code wrong?
2) Is it my understanding of the proper process that is wrong?
3) Have I missing something?

Thanks

**Perfumes.php**
``` php
public $selected = [];

protected $rules = [
     "selected.*" => "required|integer"
];

public function updatedSelected( $values )
{
    $this->validate();

     if( $this->plant->id != null )
     {
         $this->plant->perfumes()->sync(
             array_map( "intval", $values )
         );
     }
}
```

**perfumes.blade.php**
``` php
<div class="auto-grid-placement auto-grid-item">
    @foreach( $perfumes as $perfume )
        @php
            $slug = slugafy( $perfume->name );
        @endphp
        <x-form.label for="{{ $slug }}-{{ $perfume->id }}" class="{{ $disabled }} checkbox-item px-2 py-2 inline-block">
            <input wire:model.lazy="selected" type="checkbox" id="{{ $slug }}-{{ $perfume->id }}" class="" value="{{ $perfume->id }}" {{ $disabled }}> {{ $perfume->name }} - {{ $perfume->id }}
        </x-form.label>
    @endforeach
</div>
```

**Comment by jonneroelofs** [comment]
Are you sure that it is passing your validation? Because you declare $selected as an array, but you validate that it is in integer. Maybe change it to something like:

```php
protected $rules = [
     "selected" => ['required', 'array'],
     "selected.*" => ['integer']
];
```

**Reply by marcdegagne** [reply]
Thanks for your reply @[jonneroelofs](https://github.com/jonneroelofs).
Yes it passes validation. I dd() the $value in the updatedSelected() method after validate().
Values do get saved.
It's when checkboxes get unchecked, the array returned is the same as when the page was initially loaded.
Obviously your validation is properly more accurate.
I tested your suggestion and same result.

**Comment by joshhanley** [comment]
@MDGDesign what patch version of Livewire are you using?

Your setup looks like it should work. Have a look at this free screencast we have on checkboxes which may help https://laravel-livewire.com/screencasts/form-checkboxes

Hope this helps!

**Reply by marcdegagne** [reply]
@joshhanley I am using 2.8.0.
Thanks for the reminder, I did watch it before I started the checkboxes.
I don't know yet how, but I will try to get if the checkbox is checked.
I will try also on an isolated project and recreate the component alone to sort out outside influences.

Thanks

**Comment by marcdegagne** [comment, marked as answer]
Finally...

I was able to get through this by changing the ```wire:model="selected"``` to ```wire:click=”savePerfumes( $perfume->id)”``` by calling a method to filter and save the input values. This is the example that got me on the right track: https://www.youtube.com/watch?v=gNuWLHiRn-o

I must then conclude that the approach of calling the ```wire:model=”selectedPerfume”```, the set array, on every checkbox is probably better, if not made for, single-use checkboxes and these are not checked from values from the DB on load, like most examples you see.

If one needs to check and uncheck checkboxes, this is my approach:

**The view**
```php
<div class="auto-grid-placement auto-grid-item">
    
    @foreach( $perfumes as $perfume )
        @php
            $slug = slugafy( $perfume->name );
            $checked = ( in_array( $perfume->id, $selectedPerfumes->toArray() )? "checked='checked'" : "")
        @endphp
      
        <x-form.label for="{{ $slug }}-{{ $perfume->id }}" class="{{ $disabled }} checkbox-item px-2 py-2 inline-block">
            <input wire:click="savePerfumes( {{ $perfume->id }} )" type="checkbox" id="{{ $slug }}-{{ $perfume->id }}" name="selectedPerfumes[]" {{ $checked }} class="" value="{{ $perfume->id }}" {{ $disabled }}> {{ $perfume->name }} - {{ $perfume->id }}
        </x-form.label>
    @endforeach
  
</div>
```

**The called function on click**
```php

// The new $rules
 protected $rules = [
        "selectedPerfumes"   => "array",
        "selectedPerfumes.*" => "required|integer"
    ];

 public function savePerfumes( $perfume_id )
    {
        $this->validate();

        if( $this->selectedPerfumes->contains( $perfume_id ))
        {
            // Removing
            $this->selectedPerfumes = $this->selectedPerfumes->reject( function( $value ) use ($perfume_id) {
                return $value == $perfume_id;
            });
        }else{
             // Adding
            $this->selectedPerfumes->push( $perfume_id );
        }

        if( $this->plant->id != null )
        {
            $this->plant->perfumes()->sync(
                $this->selectedPerfumes
            );
        }
    }
```
